### PR TITLE
feat(prejoin) expose prejoin page

### PR DESCRIPTION
**Pull request by quitrk**
Added capability for rendering different app screens from the same app.bundle.
Exposed Prejoin screen to allow reuse of its video preview and device selection capabilies

### CODE REVIEW

**Comment by quitrk** [comment]
@saghul

**Comment by beyondlevi** [comment]
This is amazing. Please, release that.

File: index.html [hunk 1]
```diff
@@ -8,7 +8,13 @@
 
     <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
     <link rel="stylesheet" href="css/all.css">
-
+    <script>
+        document.addEventListener('DOMContentLoaded', () => {
```
Comment (saghul) [inline index.html:12]: Won't this also trigger even when we failed to load some of the critical files?

File: react/features/base/util/parseURLParams.js [hunk 2]
```diff
@@ -17,7 +17,7 @@ export function parseURLParams(
         url: URL,
         dontParse: boolean = false,
         source: string = 'hash'): Object {
-    const paramStr = source === 'search' ? url.search : url.hash;
+    const paramStr = url && (source === 'search' ? url.search : url.hash);
```
Comment (saghul) [inline react/features/base/util/parseURLParams.js:20]: how come this is called without an actual URL?

File: react/features/prejoin/components/PrejoinApp.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     * Renders the platform specific dialog container.
+     *
+     * @returns {React$Element}
+     */
+    _renderDialogContainer() {
+        return (
+            <AtlasKitThemeProvider mode = 'dark'>
+                <DialogContainer />
```
Comment (saghul) [inline react/features/prejoin/components/PrejoinApp.js:158]: What dialogs do we need to display here?

File: react/features/prejoin/reducer.js [hunk 4]
```diff
@@ -29,11 +29,23 @@ const DEFAULT_STATE = {
     userSelectedSkipPrejoin: false
 };
 
+const STORE_NAME = 'features/prejoin';
+
+/**
+ * Sets up the persistence of the feature {@code prejoin}.
+ */
+const filterSubtree = {
```
Comment (saghul) [inline react/features/prejoin/reducer.js:37]: These are no longer stored in the prejoin feature. Are you sure you are running latest master as the base for your PR?

File: static/prejoin.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                showAvatar,
+                showJoinActions
+              }
+          })
+      })
+    </script>
+    <!--#include virtual="/title.html" -->
+    <script><!--#include virtual="/config.js" --></script><!-- adapt to your needs, i.e. set hosts and bosh path -->
+    <script><!--#include virtual="/interface_config.js" --></script>
+    <script src="../libs/lib-jitsi-meet.min.js?v=139"></script>
```
Comment (saghul) [inline static/prejoin.html:31]: @damencho / @aaronkvanmeerten Do we need to do something so these are automagically updated when needed?

File: react/features/prejoin/middleware.js [hunk 6]
```diff
@@ -29,8 +31,26 @@ MiddlewareRegistry.register(store => next => async action => {
 
         break;
     }
+
+    case SET_AUDIO_MUTED: {
```
Comment (saghul) [inline react/features/prejoin/middleware.js:35]: This change could go in its own commit, since it's not really related to exposing the Prejoin screen as an app.

File: react/features/prejoin/components/PrejoinApp.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                }
+            });
+
+            dispatch(setConfig({ prejoinPageEnabled: true }));
+
+            await dispatch(configureInitialDevices());
+            await dispatch(getAvailableDevices());
+
+            const { tryCreateLocalTracks, errors } = conference.createInitialLocalTracks(
+                conference._getInitialOptions()
```
Comment (saghul) [inline react/features/prejoin/components/PrejoinApp.js:127]: I'd like us to not use `conference.js` here at all. Even if that means duplicating a bit of code. That file is legacy and the less we add to it or depend on it the better.

File: css/_premeeting-screens.scss [hunk 8]
```diff
@@ -1,22 +1,30 @@
 /**
  * Shared style for full screen local track based dialogs/modals.
  */
+ .premeeting-screen,
```
Comment (saghul) [inline css/_premeeting-screens.scss:4]: @vp8x8 Can you please review the CSS here, and the changes to the Prejoin component? The latter LGTM, but an extra pair of eyes won't hurt.

Comment (vp8x8) [reply css/_premeeting-screens.scss:4]: LGTM

File: react/features/base/util/parseURLParams.js (hunk 2 above)
Comment (quitrk) [reply react/features/base/util/parseURLParams.js:20]: `function getDevicesFromURL(state: Object) {
    const urlParams
        = parseURLParams(state['features/base/connection'].locationURL);`

This is called when configuring the initial devices. Since there's no connection for this case, there's also no URL

File: index.html (hunk 1 above)
Comment (quitrk) [reply index.html:12]: I will add a guard for the truthiness of JitsiMeetJS, but wouldn't we have bigger problems if that happens than some exception being thrown in the console?

File: react/features/prejoin/components/PrejoinApp.js (hunk 3 above)
Comment (quitrk) [reply react/features/prejoin/components/PrejoinApp.js:158]: None, will change to return null since the function still needs to be implemented

File: react/features/prejoin/reducer.js (hunk 4 above)
Comment (quitrk) [reply react/features/prejoin/reducer.js:37]: Missed that change when rebased, updating

File: react/features/base/tracks/functions.js [hunk 9]
```diff
@@ -125,6 +126,105 @@ export function createLocalTracksF(options = {}, firePermissionPromptIsShownEven
         }));
 }
 
+/**
+ * Returns an object containing a promise which resolves with the created tracks &
+ * the errors resulting from that process.
+ *
+ * @returns {Promise<JitsiLocalTrack>}
+ */
+export function createPrejoinTracks() {
```
Comment (quitrk) [inline react/features/base/tracks/functions.js:135]: moved here implementation from conference.js, with the exception of the screen sharing handling

File: react/features/prejoin/middleware.js (hunk 6 above)
Comment (quitrk) [reply react/features/prejoin/middleware.js:35]: moved it into a separate commit, but kept it in the same pull request

File: react/features/prejoin/components/PrejoinApp.js (hunk 7 above)
Comment (quitrk) [reply react/features/prejoin/components/PrejoinApp.js]: done

File: react/features/base/util/parseURLParams.js (hunk 2 above)
Comment (saghul) [reply react/features/base/util/parseURLParams.js:20]: Then can we perhaps use a different function, since there is no URL really, so we probably shouldn't be trying to load things from there.

File: index.html (hunk 1 above)
Comment (saghul) [reply index.html:12]: We have that inline HTML page telling you things are wrong, in principle things would stop there.

File: react/features/base/tracks/middleware.js [hunk 10]
```diff
@@ -143,6 +143,10 @@ MiddlewareRegistry.register(store => next => action => {
         if (typeof APP !== 'undefined') {
             const result = next(action);
 
+            if (isPrejoinPageVisible(store.getState())) {
```
Comment (saghul) [inline react/features/base/tracks/middleware.js:146]: why is this change necessary?

File: react/features/base/util/parseURLParams.js (hunk 2 above)
Comment (quitrk) [reply react/features/base/util/parseURLParams.js:20]: removed the parent call completely since it turns out it wasn't required

File: react/features/base/tracks/middleware.js (hunk 10 above)
Comment (quitrk) [reply react/features/base/tracks/middleware.js:146]: not applicable anymore after removing the call to `getAvailableDevices`

File: index.html [hunk 11]
```diff
@@ -8,7 +8,17 @@
...
     <link rel="stylesheet" href="css/all.css">
+    <script>
+        document.addEventListener('DOMContentLoaded', () => {
+            if (!JitsiMeetJS.app) {
+                return;
+            }
 
+            JitsiMeetJS.app.renderEntryPoint({
+                Component: JitsiMeetJS.app.entryPoints.APP
+            })
```
Comment (saghul) [inline index.html:19]:
```suggestion
            });
```

File: react/features/prejoin/components/PrejoinApp.js [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+import { initPrejoin } from '../actions';
+
+import Prejoin from './Prejoin';
+
+type Props = {
+
+    /**
+     * Flag signaling if the device status is visible or not.
+     */
+    deviceStatusVisible: boolean,
```
Comment (saghul) [inline react/features/prejoin/components/PrejoinApp.js:18]: Aren't most of these props unused ?
